src/fromager/commands/bootstrap.py

```diff
@@ -82,24 +85,125 @@ def bootstrap(
     with open(wkctx.work_dir / "constraints.txt", "w") as f:
-        for name, version in sorted(
-            set(
-                _installation_dependencies(
-                    wkctx.all_edges, context.ROOT_BUILD_REQUIREMENT
-                )
-            )
-        ):
-            f.write(f"{name}=={version}\n")
+        write_constraints_file(graph=wkctx.all_edges, output=f)
 
 
-def _installation_dependencies(
-    # The keys of all_edges will be either a validate package name or an empty
-    # string.
+def write_constraints_file(
+    graph: context.BuildRequirements,
+    output: typing.TextIO,
+) -> None:
+    reverse_graph: ReverseRequirements = reverse_dependency_graph(graph)
+
+    # The installation dependencies are (name, version) pairs, and there may be
+    # duplicates because multiple packages depend on the same version of another
+    # package. Eliminate those duplicates using a set, so we can more easily
+    # find cases where we depend on two different versions of the same thing.
+    install_constraints = set(
+        (name, version)
+        for name, version, _ in installation_dependencies(
+            all_edges=graph,
+            name=context.ROOT_BUILD_REQUIREMENT,
+            version=None,
+        )
+    )
+
+    # Look for potential conflicts by tracking how many different versions of
+    # each package are needed.
+    conflicts: dict[NormalizedName, list[Version]] = {}
+    for dep_name, dep_version in install_constraints:
+        conflicts.setdefault(dep_name, []).append(dep_version)
+
+    for dep_name, versions in sorted(conflicts.items()):
+        if len(versions) == 0:
+            # This should never happen.
+            raise ValueError(f"No versions of {dep_name} supported")
+
+        if len(versions) == 1:
+            # This is going to be the situation for most dependencies, where we
+            # only have one version.
+            output.write(f"{dep_name}=={versions[0]}\n")
+            continue
+
+        # Below this point we have built multiple versions of the same thing, so
+        # we need to try to determine if any one of those versions meets all of
+        # the requirements.
+
+        # Track which versions can be used by which parent requirement.
+        usable_versions: dict[str, list[str]] = {}
+        # Track how many total users of a requirement (by name) there are so we
+        # can tell later if any version can be used by all of them.
+        user_counter = 0
+
+        # Which parent requirements can use which versions of the dependency we
+        # are working on?
+        for dep_version in sorted(versions):
+            key = f"{dep_name}=={dep_version}"
+            parent_info = reverse_graph.get(key, [])
+            user_counter += len(parent_info)
+            for parent_version, req in parent_info:
+                match_versions = [str(v) for v in versions if v in req.specifier]
```

Comment:
I think we can use `filter` here: https://packaging.pypa.io/en/stable/specifiers.html#packaging.specifiers.SpecifierSet.filter